DESIGN_DOC.md

```diff
@@ -0,0 +1,99 @@
+
+If the first method is selected, we will need to determine the precise coordinates of the calendar event since the calendar API only provides a textual representation of the address. We can use Google's Maps API to achieve that. We can also use this API to determine the distance between two points on the globe.
+
+If we choose the second approach, we would still need to use the Maps API to determine the country of the address. However, this time it would be more complicated to determine if two events are from the same location since we can have two different strings describing the same country (e.g. `USA` and `US` or `US` and `United States of America`). Thus, I think that the first approach is the better option.
+
+### User-specified configurations
+It might be a good idea to allow users to configure various parameters of our API. These might include:
+1. The time interval for the Calendar events that can receive warnings from our API.
+Users can create calendar events for any point in time. It doesn't make any sense to produce warnings for events from the past since those have already happened. It also doesn't make much sense to produce warnings for the events in the far future since a lot can change before those events occur. Thus, the user can specify the amount of days to determine which calendar events should be monitored. We call this `Calendar time-frame`. More precisely, given the calendar time-frame `T` and some calendar event `E`, event `E` will only be monitored if it's time of arrival is less that or equal to `now + T`. Note that, since the user can create calendar events for arbitrary points in time, some of the created events might become "monitorable" in the future.
+2. Google calendar which we can monitor events from. I think this feature can be omitted for now and we can monitor all events from the `primary` calendar, which is available for every user by default.
+
+It might also be useful to provide an endpoint allowing users to change these settings in the future.
+
+### System design
+The simplest approach would be to use the following system:
+![simple design](simple-disaster-tracker.png)
+
+There are a few problems/possibilities for improvement with this approach:
+- The amount of time it takes to produce a warning for a newly created calendar event depends on the scheduling settings of the `Warning producer` cron job and the `Disaster fetcher` cron job.
+- `Warning producer` can be quite inefficient since it will have to:
+  - fetch all the disasters and all the calendar events
+  - cross check them with each other and produce warnings
+  - take into account that different users have different calendar time-frames.
+It's probably possible, though, to fetch all monitored calendar events using a single SQL query with joins, so this shouldn't be a problem.
+
```

Comment:
I think going with this design is the most reasonable choice. Only change I would make is adding a message queue (SQS or Kafka) between warning producer and warning manager. This will help because:
1. In future, we can have many warning producers (e.g. prioritized, non-prioritized), and number of events can get big. Therefore sending directly to the warning manager may overload him. Having a queue in between and allowing them to accumulate in queue is ok.
2. Ability to scale managers independently (e.g. if we have millions of devices and tens of millions of notifs). We may need to scale the warning producers, therefore using queue can allow us easily increase number of such servers.
3. Ability to indirectly send warnings to the queue (e.g. if there is some event not provided via disaster API, but we know about it). We can send events to the users by hand (or script), even if it is not passed via disaster API. **This last point is questionable, there could be better implementations if there will be human monitoring of disasters.**